### PR TITLE
[FLINK-20539][table-planner] fix type mismatch when using ROW in computed column

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

When using row as a computed column and select the computed column as a query, an exception will be thrown by calcite because of the type mismatch about ROW.

The cause is that in Flink, we always treat the ROW as a struct type PEEK_FIELDS_NO_EXPAND (refs to LogicalRelDataTypeConverter and FlinkTypeFactory), but in calcite, the struct type about ROW is FULLY_QUALIFIED (refs to SqlRowOperator#inferReturnType). And when selecting the column, the difference causes this bug.

## Brief change log

  - *Modify the SqlRowOperator to return the true struct type about ROW*
  - *Add tests about this pr*

## Verifying this change

Necessary tests have been added.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Review by LadyForest — APPROVED** [review]
Thanks for the contribution! The fix looks good to me.

**Comment by xuyangzhong** [comment]
Hi, @snuyanzin, sorry for this noise. I noticed that you participated in some changes about Calcite in  Flink before. I would be very grateful if you had time to review this PR for me.

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlRowOperator.java [hunk 1]
```diff
@@ -66,12 +70,17 @@
  *       </ul>
  * </ul>
  *
- * Once Flink applies same logic for both table api and sql, this class should be removed.
+ * <p>Once Flink applies same logic for both table api and sql, this first changes should be
+ * removed.
+ *
+ * <p>2. It uses PEEK_FIELDS_NO_EXPAND with a nested struct type (Flink [[RowType]]).
+ *
+ * <p>See more at {@code LogicalRelDataTypeConverter} and {@code FlinkTypeFactory}.
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlRowOperator.java:78]:
```suggestion
 * <p>See more at {@link LogicalRelDataTypeConverter} and {@link FlinkTypeFactory}.
```
i guess it's better to use `@link` here

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlRowOperator.java [hunk 2]
```diff
@@ -66,12 +70,17 @@
  *       </ul>
  * </ul>
  *
- * Once Flink applies same logic for both table api and sql, this class should be removed.
+ * <p>Once Flink applies same logic for both table api and sql, this first changes should be
+ * removed.
+ *
+ * <p>2. It uses PEEK_FIELDS_NO_EXPAND with a nested struct type (Flink [[RowType]]).
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlRowOperator.java:76]:
```suggestion
 * <p>2. It uses {@link PEEK_FIELDS_NO_EXPAND} with a nested struct type (Flink [[{@link RowType}]]).
```
i wonder whether usage of links could improve navigation in documentation

**Comment by snuyanzin** [comment]
thanks for the contribution
in general it looks ok from my side
i left a couple of minor comments

**Comment by xuyangzhong** [comment]
Hi, @snuyanzin . Thanks for your review. 

The bug happens only when using the SQL API. The table API will not throw this exception actually. The quick explanation is: 

- Only when parsing SQL, the struct kind of the computed ROW type will be changed to PEEK_FIELDS_NO_EXPAND by calling FlinkTypeFactory#createFieldTypeFromLogicalType after calling SqlRowOperator.
- When using table API, the struct kind of the computed ROW type is always FULLY_QUALIFIED by only calling SqlRowOperator.

The following is the detail:

After executing DDL 'create table', the struct kind of the computed ROW type is FULLY_QUALIFIED (which is wrong across SqlRowOperator), but Flink always replaces the Calcite RelDataType Row with Flink LogicalType Row and ignores the struct kind as the source schema. 

When executing the query 'select * from table', Flink will get the source output RelDataType in CatalogSchemaTable#getRowType, where Flink reverts the LogicalType to calcite RelDataType and the RelDataType ROW in Calcite is attached with PEEK_FIELDS_NO_EXPAND in FlinkTypeFactory#createFieldTypeFromLogicalType. However, in source columns, the output RelDataType of the expression "row(a, b)" is still with FULLY_QUALIFIED, which causes the validation to fail later.

The table API will not throw this exception. Although table API shares the code about converting the "row(a, b)" to expression and gets the wrong FULLY_QUALIFIED when building source, it lakes the piece about reverting Flink LogicalType to calcite RelDataType in CatalogSchemaTable#getRowType(This is necessary for Calcite parsing & validating) and using the wrong FULLY_QUALIFIED as the final struct kind about Row.

I think for fixing this bug, this pr is already. But for more, we need to union the behavior about these two APIs.

**Comment by xuyangzhong** [comment]
@snuyanzin Hi, could you do the final check about this pr if you have time?

**Review by snuyanzin — APPROVED** [review]

**Comment by JingGe** [comment]
@xuyangzhong Thanks for driving it! NIT: remove `(for example:)` in PR description. It belongs to PR template only.

**Review by JingGe — APPROVED** [review]
LGTM

**Comment by LadyForest** [comment]
@xuyangzhong Could you help to pick the fix to release-1.18?

**Comment by xuyangzhong** [comment]
> @xuyangzhong Could you help to pick the fix to release-1.18?

Sure, this is the pr cherry-picking to 1.18. https://github.com/apache/flink/pull/23772

**Pull request merged**